translator/trace/protospan_to_jaegerthrift.go

```diff
@@ -0,0 +1,321 @@
+			jRefType = jaeger.SpanRefType_CHILD_OF
+		default:
+			// TODO: (@pjanotti) Jaeger doesn't have a unknown SpanRefType, it has FOLLOWS_FROM or CHILD_OF
+			// at first mapping all others to FOLLOWS_FROM.
+			jRefType = jaeger.SpanRefType_FOLLOWS_FROM
+		}
+
+		jRef := &jaeger.SpanRef{
+			RefType: jRefType,
+			SpanId:  ocIDBytesToJaegerID(ocLink.SpanId),
+		}
+		jRef.TraceIdLow, jRef.TraceIdHigh = traceIDBytesToLowAndHigh(ocLink.TraceId)
+		jRefs = append(jRefs, jRef)
+	}
+
+	return jRefs
+}
+
+func appendJaegerTagFromOCSpanKind(jTags []*jaeger.Tag, ocSpanKind tracepb.Span_SpanKind) []*jaeger.Tag {
+	// We could check if the key is already present but it doesn't seem worth at this point.
+	// TODO: (@pjanotti): Replace any OpenTracing literals by importing github.com/opentracing/opentracing-go/ext?
+	var tagValue string
+	switch ocSpanKind {
+	case tracepb.Span_CLIENT:
+		tagValue = "client"
+	case tracepb.Span_SERVER:
+		tagValue = "server"
+	}
+
+	if tagValue != "" {
+		jTag := &jaeger.Tag{
+			Key:  "span.kind",
+			VStr: &tagValue,
+		}
+		jTags = append(jTags, jTag)
+	}
+
+	return jTags
+}
+
+func ocTimeEventsToJaegerLogs(ocSpanTimeEvents *tracepb.Span_TimeEvents) []*jaeger.Log {
+	if ocSpanTimeEvents == nil || ocSpanTimeEvents.TimeEvent == nil {
+		return nil
+	}
+
+	ocTimeEvents := ocSpanTimeEvents.TimeEvent
+
+	// Assume that in general no time events are going to produce nil Jaeger logs.
+	jLogs := make([]*jaeger.Log, 0, len(ocTimeEvents))
+	for _, ocTimeEvent := range ocTimeEvents {
+		jLog := &jaeger.Log{
+			Timestamp: timestampToEpochMicroseconds(ocTimeEvent.Time),
+		}
+		switch teValue := ocTimeEvent.Value.(type) {
+		case *tracepb.Span_TimeEvent_Annotation_:
+			jLog.Fields = ocAnnotationToJagerTags(teValue.Annotation)
+		case *tracepb.Span_TimeEvent_MessageEvent_:
+			jLog.Fields = ocMessageEventToJaegerTags(teValue.MessageEvent)
+		default:
+			msg := "An unknown OpenCensus TimeEvent type was detected when translating to Jaeger"
+			jTag := &jaeger.Tag{
+				Key:  "unknown.oc.timeevent.type",
+				VStr: &msg,
+			}
+			jLog.Fields = append(jLog.Fields, jTag)
+		}
+
+		jLogs = append(jLogs, jLog)
+	}
+
+	return jLogs
+}
+
+func ocAnnotationToJagerTags(annotation *tracepb.Span_TimeEvent_Annotation) []*jaeger.Tag {
+	if annotation == nil {
+		return nil
+	}
+
+	// TODO: (@pjanotti) what about Description? Does it fit as another tag?
+
+	return ocSpanAttributesToJaegerTags(annotation.Attributes)
+}
+
+func ocMessageEventToJaegerTags(msgEvent *tracepb.Span_TimeEvent_MessageEvent) []*jaeger.Tag {
+	if msgEvent == nil {
+		return nil
+	}
+
+	// TODO: (@pjanotti) Not clear how to map those to Jaeger, perhaps some OpenTracing tags...
+
+	return nil
+}
+
+func truncableStringToStr(ts *tracepb.TruncatableString) string {
+	if ts == nil {
+		return ""
+	}
+	return ts.Value
+}
+
+func traceIDBytesToLowAndHigh(traceID []byte) (traceIDLow, traceIDHigh int64) {
+	if len(traceID) != 16 {
+		return
```

Comment:
It would be useful to return an error instead of silently passing as TraceID is a very important field.